runtime/src/main/cpp/Memory.cpp

```diff
@@ -639,11 +640,47 @@ inline void initThreshold(MemoryState* state, uint32_t gcThreshold) {
 }
 #endif // USE_GC
 
-#if TRACE_MEMORY || USE_GC
+#if TRACE_MEMORY && USE_GC
+
+const char* colorNames[] = {"BLACK", "GRAY", "WHITE", "PURPLE", "GREEN", "ORANGE", "RED"};
+
+void dumpObject(ObjHeader* ref, int indent) {
+  for (int i = 0; i < indent; i++) MEMORY_LOG(" ");
+  auto* typeInfo = ref->type_info();
+  auto* packageName =
+    typeInfo->packageName_ != nullptr ? CreateCStringFromString(typeInfo->packageName_) : nullptr;
+  auto* relativeName =
+    typeInfo->relativeName_ != nullptr ? CreateCStringFromString(typeInfo->relativeName_) : nullptr;
+  MEMORY_LOG("%p %s.%s\n", ref,
+    packageName ? packageName : "<unknown>", relativeName ? relativeName : "<unknown>");
+  if (packageName) konan::free(packageName);
+  if (relativeName) konan::free(relativeName);
+}
+
+void dumpContainerContent(ContainerHeader* container) {
+  if (isAggregatingFrozenContainer(container)) {
+    MEMORY_LOG("%s aggregating container %p with %d objects rc=%d\n",
+               colorNames[container->color()], container, container->objectCount(), container->refCount());
+    ContainerHeader** subContainer = reinterpret_cast<ContainerHeader**>(container + 1);
+    for (int i = 0; i < container->objectCount(); ++i) {
+      ObjHeader* obj = reinterpret_cast<ObjHeader*>(subContainer + 1);
```

Comment:
It this correct?

Reply:
Why not, supercontainer contains references to frozen objects containers.

Reply:
`obj` seems to be the same on all loop iterations, doesn't it?
Is it really located at `container + sizeof(ContainerHeader) + sizeof(ContainerHeader*)`?